File: src/se/fredsberg/game/zombiedice/Turn.java
package se.fredsberg.game.zombiedice;

import java.util.List;
import java.util.Map;

import se.fredsberg.game.container.Pool;
import se.fredsberg.game.dice.zombie.ZombieDie;
import se.fredsberg.game.dice.zombie.ZombieValue;
import se.fredsberg.game.random.Scrambler;

public class Turn {

    private Pool<ZombieDie> cup;
    private int overflowBrains = 0;
    private Map<ZombieValue, Pool<ZombieDie>> piles;
    private boolean stopped = false;

    public Turn(Pool<ZombieDie> cup, Map<ZombieValue, Pool<ZombieDie>> piles) {
        this.cup = cup;
        this.piles = piles;
    }

    public int count(ZombieValue type) {
        return piles.get(type).getCount();
    }

    public boolean itsOver() {
        return stopped || shoot();
    }

    private boolean shoot() {
        return count(ZombieValue.SHOOTGUN) >= 3;
    }

    public void keepGoing(Scrambler scrambler) {
        assertStillGoingStrong();
        for (ZombieDie die : getHandOfDice(scrambler)) {
            die.roll(scrambler);
            piles.get(die.getResult()).add(die);
        }
    }

    public int score() {
        if (shoot()) {
            return 0;
        }
        return count(ZombieValue.BRAIN) + overflowBrains;
    }

    public int stop() {
        stopped = true;
        return score();
    }

    private void assertStillGoingStrong() {
        if (itsOver()) {
            throw new IllegalStateException();
        }
    }

    private ZombieDie getDie(Scrambler scrambler) {
        if (cup.isEmpty()) {
            overflowBrains += count(ZombieValue.BRAIN);
            resetDeck(scrambler);
        }
        return cup.pickOne();
    }

    private List<ZombieDie> getHandOfDice(Scrambler scrambler) {
        List<ZombieDie> hand = piles.get(ZombieValue.FOOTPRINTS).pickAll();
        while (hand.size() < 3) {
            hand.add(getDie(scrambler));
        }
        return hand;
    }

    private void resetDeck(Scrambler scrambler) {
        for (ZombieDie die : piles.get(ZombieValue.BRAIN).pickAll()) {
            cup.add(die);
        }
        cup.shake(scrambler);
    }
}

File: jssrc/Turn.js
function Turn(cup, piles) {
	this.cup = cup;
	this.piles = piles;
	this.overflowBrains = 0;
	this.stopped = false;

	this.getHandOfDice = function(scrambler) {
		var hand = new Array(3);
		for ( var i = 0; i < hand.length; i++) {
			hand[i] = this.getDie(scrambler);
		}
		return hand;
	};

	this.getDie = function(scrambler) {
		if (this.cup.isEmpty()) {
			this.overflowBrains += this.count(ZombieValue.values.BRAIN);
			this.resetDeck(scrambler);
		}

		return this.cup.pickOne();
	};

	this.resetDeck = function(scrambler) {
		var brains = piles[ZombieValue.values.BRAIN].pickAll();
		for ( var i = 0; i < brains.length; i++) {
			this.cup.add(brains[i]);
		}
		this.cup.shake(scrambler);
	};

	this.shoot = function() {
		return this.count(ZombieValue.values.SHOOTGUN) >= 3;
	};
	
	this.stillGoingStrong = function() {
		return !this.itsOver();
	}
};

Turn.prototype.count = function(zombieValue) {
	return this.piles[zombieValue].getCount();
};

Turn.prototype.keepGoing = function(scrambler) {
	if (this.stillGoingStrong()) {
		var handOfDice = this.getHandOfDice(scrambler);
		for ( var i = 0; i < handOfDice.length; i++) {
			var die = handOfDice[i];
			die.roll(scrambler);
			this.piles[die.getResult()].add(die);
		}
	}
};

Turn.prototype.score = function() {
	if (this.shoot()) {
		return 0;
	}
	return this.count(ZombieValue.values.BRAIN) + this.overflowBrains;
};

Turn.prototype.itsOver = function() {
	return this.stopped || this.shoot();
};
